localstack/utils/net.py

```diff
@@ -193,17 +201,74 @@ def get_free_udp_port(blocklist: List[int] = None) -> int:
 
 
 def get_free_tcp_port(blocklist: List[int] = None) -> int:
+    """
+    Tries to bind a socket to port 0 and returns the port that was assigned by the system. If the port is
+    in the given ``blocklist``, or the port is marked as reserved in ``dynamic_port_range``, the procedure
+    is repeated for up to 50 times.
+
+    :param blocklist: an optional list of ports that are not allowed as random ports
+    :return: a free TCP port
+    """
     blocklist = blocklist or []
-    for i in range(10):
+    for i in range(50):
         tcp = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
         tcp.bind(("", 0))
         addr, port = tcp.getsockname()
         tcp.close()
-        if port not in blocklist:
+        if port not in blocklist and not dynamic_port_range.is_port_reserved(port):
+            try:
+                dynamic_port_range.mark_reserved(port)
+            except ValueError:
+                # depending on the ephemeral port range of the system, the allocated port may be outside what
+                # we defined as dynamic port range
+                pass
             return port
     raise Exception(f"Unable to determine free TCP port with blocklist {blocklist}")
 
 
+def get_free_tcp_port_range(num_ports: int, max_attempts: int = 50) -> "PortRange":
+    """
+    Attempts to get a contiguous range of free ports from the dynamic port range. For instance,
+    ``get_free_tcp_port_range(4)`` may return the following result: ``PortRange(44000:44004)``.
+
+    :param num_ports: the number of ports in the range
+    :param max_attempts: the number of times to retry if a contiguous range was not found
+    :return: a port range of free TCP ports
+    :raises PortNotAvailableException: if max_attempts was reached to re-try
+    """
+    if num_ports < 2:
+        raise ValueError(f"invalid number of ports {num_ports}")
+
+    def _is_port_range_free(_range: PortRange):
+        for _port in _range:
+            if dynamic_port_range.is_port_reserved(_port) or not port_can_be_bound(_port):
+                return False
+        return True
+
+    attempts = 0
+    while attempts <= max_attempts:
+        attempts += 1
```

Comment:
nit:
```suggestion
    for _ in range(max_attempts):
```

Reply:
That would change the semantics though, as we would now retry it `max_attempts + 1` times.
Since max attempts is defined as `    :param max_attempts: the number of times to retry if a contiguous range was not found`, I would argue it should the +1 behavior is correct, even though we might want to rename the "attempts" part of the parameter.

Reply:
with `for _ in range(max_attempts)` there will be exactly 50 calls to `port_can_be_bound` (supposing each one fails). i think that's fine :-) added the check to the test